### PR TITLE
Openfam v3.0

### PR DESCRIPTION
**Pull request by BA-Ramya**
OpenFAM 3.0 documentation changes

### CODE REVIEW

File: release_notes.html [hunk 1]
```diff
@@ -89,23 +95,43 @@ <h1>OpenFAM Reference Implementation</h1>
...
+			<li>Multi-threaded support has been added that enables OpenFAM to be used with multi-threaded applications. See <a href="multithread.html">multi-threaded environment</a> for more details.</li>
+			<li>Performance improvements:
+				Data item interleaving feature has been added in which a given data item will span across multiple memory servers. This has been done to uniformly distribute the load across multiple memory servers, thus reducing the network bottlenecks, and thereby improving the overall performance. Data item interleaving is enabled by default. See  <a href="interleaving.html">Data Item Interleaving </a> for more details.</li>
+			<li> Added support for multiple contexts. The data path APIs (get,put, scatter, gather and atomics) can be invoked from user defined contexts. See <a href="fam_context_open.html"> fam_context_open API </a> for more details.</li>
+			<li>		OpenFAM APIs changes:
+		<ul>
+		<li>fam_create_region API has been modified wherein the input argument redundancyLevel is replaced with a structure that holds attributes for a region namely - redundancyLevel, memoryType and flags to indicate if data item interleaving needs to be enabled or not.</li>
```
Comment (ssinghal53) [inline release_notes.html]: <code>fam_create_region</code> ... <code>redundancyLevel</code> etc.

File: v3.0/release_notes.html [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+		<ul>
+		<li>fam_create_region API has been modified wherein the input argument redundancyLevel is replaced with a structure that holds attributes for a region namely - redundancyLevel, memoryType and flags to indicate if data item interleaving needs to be enabled or not.</li>
+		<li>Two new APIs have been added to support multiple contexts namely <code>fam_context_open()</code> and <code>fam_context_close()</code>. These APIs provide the flexibility to threads to invoke datapath operations parallelly. </li>
+		<li>A new API <code>fam_progress</code> has been provided that returns the count of pending datapath and atomic operations.</li>
+		<li>Few APIs have been added to enable backing up and restoring of data items namely - <code>fam_backup()</code> and <code>fam_restore()</code>. An API has been provided to delete the unused backup in the memory server.</li>
+		</ul>
+		</ul>
+		
+		<h2>OpenFAM version 2.0 (release v2.0.0)</h2>
+		<p>The current version of OpenFAM is 2.0. It contains the following changes and enhancements over version 1.0:</p>
```
Comment (ssinghal53) [inline v3.0/release_notes.html]: Change to "OpenFAM 2.0 contains the following ..."

File: v3.0/release_notes.html [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+			<li> Added support for multiple contexts. The data path APIs (get,put, scatter, gather and atomics) can be invoked from user defined contexts. See <a href="fam_context_open.html"> fam_context_open API </a> for more details.</li>
+			<li>		OpenFAM APIs changes:
+		<ul>
+		<li>fam_create_region API has been modified wherein the input argument redundancyLevel is replaced with a structure that holds attributes for a region namely - redundancyLevel, memoryType and flags to indicate if data item interleaving needs to be enabled or not.</li>
+		<li>Two new APIs have been added to support multiple contexts namely <code>fam_context_open()</code> and <code>fam_context_close()</code>. These APIs provide the flexibility to threads to invoke datapath operations parallelly. </li>
```
Comment (ssinghal53) [inline v3.0/release_notes.html]: replace "parallelly." with "in parallel."

File: v3.0/limitations.html [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+</tbody>
+</table>
+As mentioned in <cite><a href="https://link.springer.com/chapter/10.1007%2F978-3-030-04918-8_5">The OpenFAM API: A Programming Model for Disaggregated Persistent Memory</a></cite>,
+the reference implementation uses libfabric
+to handle the underlying memory traffic. Libfabric uses 64 bit keys to manage memory regions
+and associated access permissions. Within OpenFAM, every data item requires independent access
+permissions, so OpenFAM data items are mapped to libfabric regions for access from remote PEs. Rather than
+assigning random keys for data items within libfabric, access keys are explicitly constructed using
+a combination of OpenFAM region Id, the data item id and OpenFAM permissions using the
+sequence Key = &lt;Region Id >&lt;DataItem Id>&lt;R|RW>. This enables the library to
```
Comment (ssinghal53) [inline v3.0/limitations.html]: "Key = &lt;Region Id >&lt;DataItem Id>&lt;R|RW>." Is this still correct?

File: v3.0/limitations.html [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+<caption>Table 3: OpenFAM APIs currently not available in the reference implementation</caption>
+<tbody>
+<tr><th>API</th><th>Memory Server</th><th>Shared Memory</th><th>Comment</th></tr>
+<tr><td class="code">fam_map</td><td>NO</td><td>YES</td><td>Libfabric currently does not support mmap APIs</td></tr>
+<tr><td class="code">fam_unmap</td><td>NO</td><td>YES</td><td>Libfabric currently does not support mmap APIs</td></tr>
+<tr><td class="code">fam_fence</td><td>YES (only with sockets provider)</td><td>NO</td><td>Fence operation is not supported by PSM2 and Verbs libfabric provider.<br/>Support for other providers and Shared Memory model will be added in future</td></tr>
+</tbody>
+</table>
+Since the underlying libfabric providers do not support 128-bit compare and swap (CAS) operations, CAS operations have been implemented
+in software.
```
Comment (ssinghal53) [inline v3.0/limitations.html:138]: Is this still correct, or do we now support 128-bit CAS?

File: v3.0/services.html [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+			</ul>
+			<hr>
+		</nav>
+		<article>
+		<h1>OpenFAM 3.0 Services</h1>
+<p> OpenFAM version 2.0 introduced the concept of supporitng large pool of memory using multiple memory servers.
+OpenFAM version 3.0 contains significant design changes to increase scalability and improve performance.
+<ul>
+<li>The design supports a larger pool of memory by supporting multiple memory servers. This memory pool can be either volatile or persistent.</li>
+<li>Data can be interleaved as dataitems can now span across multiple memory servers, thus reducing performance bottlenecks.</li>
```
Comment (ssinghal53) [inline v3.0/services.html]: Search for "dataitem"/"dataitems" throughout the documentation and replace with "data item"/"data items" (note the space between "data" and "item")

File: v3.0/services.html [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+<li>Datapath operations can now be partitioned into multiple contexts.</li>
```
Comment (ssinghal53) [inline v3.0/services.html:112]: Search for "datapath" throughout the documentation and replace with "data path" (note space between "data" and "path"

**Comment by ssinghal53** [comment]
In errors.html, second paragraph, remove 2.0 from OpenFAM reference. The description is not specific to OpenFAM 2.0

File: config_files.html [hunk 8]
```diff
@@ -218,36 +254,140 @@ <h1>Configuration Files</h1>
 <p>Sample configuration files are provided in the source under <code>openfam_parent_directory/config</code>.</p>
 <h1>Starting Services</h1>
 <p> All the services and the PE read the respective configuration file from "config" directory placed at the path mentioned
-by <code>OPENFAM_ROOT</code> environment variable or <code>/opt/OpenFAM</code>.</p>
-<p>Alternatively, most of these configuration options can be passed to services as command line arguments.
+by <code>OPENFAM_ROOT</code> environment variable or <code>By default, OPENFAM_ROOT will be /opt/hpe/OpenFAM</code>. User needs to update this environment variable to select config folder created by them</p>
+<p> A tool by name openfam_adm has been introduced which helps to perform the basic tasks useful to an OpenFAM user such as creating configuraion files, starting OpenFAM services, terminating OpenFAM services,  running tests and cleanup FAM and metadata.
```
Comment (ssinghal53) [inline config_files.html]: <code>openfam_adm</code>

File: config_files.html [hunk 9]
```diff
@@ -218,36 +254,140 @@ <h1>Configuration Files</h1>
...
+To access the tool, the openfam module must be loaded. That can be done using following command:
+
+module load openfam
```
Comment (ssinghal53) [inline config_files.html]: <pre><code>module load openfam</code></pre>

**Review by ssinghal53 — COMMENTED** [review]
In fam_initialize.html, Notes header, around line 222/261, omit "famThreadModel" from "... defaultRegionName, famThreadModel and famContextModel". OpenFAM 3.0 supports famThreadModel.

**Review by ssinghal53 — COMMENTED** [review]
fam_get_option.html: VERSION should default to 3.0

**Comment by ssinghal53** [comment]
Additional comments:

fam_destroy_region()
fan_deallocate_region()

Last sentence under Notes in both of the above: "... to perform a <code>memset</code> operation ..." (note code and insertion of "a")

fam_copy_wait occurs twice in config_files.html

fam_get_option shows !DOCTYPE HTML at the top. Should be <!DOCTYPE html> (missing "<")

fam_fetch_xor has missing menu items

**This completes my review.**

**Comment by ssinghal53** [comment]
No further comments. Everything looks good. Please merge.

**Review by mashoodk — APPROVED** [review]

**Pull request merged**